library/std/src/os/uefi/env.rs

```diff
@@ -0,0 +1,87 @@
+//! UEFI-specific extensions to the primitives in `std::env` module
+
+#![unstable(feature = "uefi_std", issue = "100499")]
+
+use crate::{cell::Cell, ffi::c_void, ptr::NonNull};
+
+// Since UEFI is single-threaded, making the global variables thread local should be safe.
+thread_local! {
+    // Flag to check if BootServices are still valid.
+    // Start with assuming that they are not available
+    static BOOT_SERVICES_FLAG: Cell<bool> = Cell::new(false);
+    // Position 0 = SystemTable
+    // Position 1 = ImageHandle
+    static GLOBALS: Cell<Option<(NonNull<c_void>, NonNull<c_void>)>> = Cell::new(None);
+}
+
+/// Initializes the global System Table and Image Handle pointers.
+///
+/// The standard library requires access to the UEFI System Table and the Application Image Handle
+/// to operate. Those are provided to UEFI Applications via their application entry point. By
+/// calling `init_globals()`, those pointers are retained by the standard library for future use.
+/// Thus this function must be called before any of the standard library services are used.
+///
+/// The pointers are never exposed to any entity outside of this application and it is guaranteed
+/// that, once the application exited, these pointers are never dereferenced again.
+///
+/// Callers are required to ensure the pointers are valid for the entire lifetime of this
+/// application. In particular, UEFI Boot Services must not be exited while an application with the
+/// standard library is loaded.
+///
+/// In normal cases, this function should only be called by Rust std internally. However, it can be
+/// called by user if they want to use Rust std without using normal Rust initialization flow. This
+/// means using attributes like `no_main` or `start` feature.
+///
+/// # SAFETY
+/// Calling this function more than once is undefined behaviour.
+pub unsafe fn init_globals(handle: NonNull<c_void>, system_table: NonNull<c_void>) {
+    GLOBALS.set(Some((system_table, handle)));
+}
```

Comment:
Shouldn't this function enable boot-services? Otherwise, it would be impossible to make use of them, since `enable_boot_services` is `pub(crate)`.

You could also check whether `ST->boot_services` is NULL. This indicates that boot-services were exited. So enable/disable them depending on whether the pointer is set?